rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,700 @@
+*   a RaggedTensor
+
+The choice depends on user’s intents, which currently is
+[implicitly](https://github.com/tensorflow/transform/blob/11afcff467f779ba6163686395582e69603987d1/tensorflow_transform/tf_metadata/schema_utils.py#L172)
+expressed in the pipeline schema.
+
+We propose to create a new [TFMD](https://github.com/tensorflow/metadata)
+(TensorFlow MetaData) Proto, `TensorRepresentation` to carry those intents implicitly:
+
+```protobuf
+message TensorRepresentation {
+  oneof {
+    DenseTensor { … }  // column_name, dtype, shape, default_value
+    VarLenSparseTensor { … }  // column_name, dtype
+    SparseTensor { }  // dtype, value_column_name, indice_column_names
+    VarLenRaggedTensor { … } // dtype
+    RaggedTensor { } // dtype, value_column_name, row_partition_column_names, ...
+    StructuredTensor { } // column_names
+  }
+}
+```
+
+This proto is used in two places:
+
+*   It’s part of TFMD schema:
+
+    ```protobuf
+      message TensorRepresentationGroup {
+        map<string, TensorRepresentation> tensor_representation = 2;
+      };
+
+      message Schema {
+       repeated Feature feature = 1;
+       // …
+       map<string, TensorRepresentationGroup> tensor_representation_group = 42;
+      }
+    ```
+
+    Note :
+
+    *   `TensorRepresentationGroup` allows different instances of one TFX
+        component to use different sets of `TensorRepresentation`s.
+    *   `tensor_representation_group` is **optional**. If the user does not
+        specify any, a default representation will be derived from
+        schema.feature to keep backwards compatibility.
+    *   this field is **not** a sub-message of Schema::Feature, because a TF
+        feedable may comprise multiple columns
+
+    Being part of the schema makes it possible to serialize and materialize the
+    intents for other components to use, which allows TFT’s materialization
+    functionality to have its own TFXIO implementation that hides the
+    data/physical format from the user.
+
+    When generating the initial schema from the statistics of the data, TFDV can
+    propose a default set of `TensorRepresentationGroup`. The user may revise
+    the proposal and TFDV can validate `TensorRepresentationGroup`s in a
+    continuous manner.
+
+*   The default implementation of TensorAdapter takes an optional `Dict[str,
+    TensorRepresentation]` at construction time. If a TFXIO implementation
+    choose to use the default TensorAdapter, it needs to provide them (may come
+    directly from the Schema).
+
+### Efficient Arrow->Tensor conversion
+
+The key to efficient conversions is to avoid copying of data. The prerequisites
+to do so are:
+
+*   Same memory alignment
+*   Same memory layout
+
+Currently 64-byte alignment is the standard in both Tensorflow's `TensorBuffer`
+and Apache Arrow's `Buffer`. Forthermore, it can be guaranteed by implementing
+our own version of `arrow::MemoryPool` that is backed by a
```

Comment:
Would this require changes in Beam to ensure that it uses our custom `MemoryPool`?

Reply:
No, at least for now, because we do decoding on our own and have full control of production of Arrow entities.